### PR TITLE
CallbackCollection

### PR DESCRIPTION
**Pull request by mosquito**

### CODE REVIEW

File: aio_pika/tools.py [hunk 1]
```diff
@@ -63,3 +64,62 @@ def wrap(*args, **kwargs):
...
+
+    def freeze(self):
+        if self.is_frozen:
+            raise RuntimeError("Collection already frozen")
+
+        self.__callbacks = frozenset(self.__callbacks)
+
+    def unfreeze(self):
+        if not self.is_frozen:
+            raise RuntimeError("Not collection is not frozen")
```
Comment (decaz) [inline aio_pika/tools.py]: Redundant "Not".

File: aio_pika/channel.py [hunk 2]
```diff
@@ -65,6 +66,14 @@ def __init__(self, connection, channel_number: int = None,
         # noinspection PyTypeChecker
         self.default_exchange = None       # type: Exchange
 
+    @property
+    def done_callbacks(self) -> CallbackCollection:
```
Comment (decaz) [inline aio_pika/channel.py]: Why do you wrap callback collections in properties?

**Comment by decaz** [comment]
@mosquito looks pretty complicated for simple storage.. Maybe it worth to leave all callback sets but make them "public" and not to embed such kind of type as `CallbackCollection` (keep it simple and let users do whatever they want with callback sets) ?

File: aio_pika/channel.py (hunk 2 above)
Comment (mosquito) [reply aio_pika/channel.py]: for restrict setter behaviour

**Comment by mosquito** [comment]
@decaz in my case freezing the callback collection it is what I really want.
The second problem is a simple set do not checks callback before add it.

File: aio_pika/tools.py [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import asyncio
 from functools import wraps
+from collections import Set
```
Comment (decaz) [inline aio_pika/tools.py]: collections.abc (to avoid `DeprecationWarning`)

**Comment by decaz** [comment]
I think something like `__call__` method could be added to `CallbackCollection` to run all the callbacks in collection instead of duplicating.

File: aio_pika/tools.py (hunk 3 above)
Comment (decaz) [reply aio_pika/tools.py]: @mosquito FYI

**Comment by decaz** [comment]
@mosquito what do you think about `CallbackCollection.__call__` method?

**Comment by mosquito** [comment]
@decaz nice idea, I will make it on separate PR

**Pull request merged**